File: test/cpu/movem.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"

#include "test.hpp"
#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;
struct F_MOVEM {
    F_MOVEM() {
        // MOVEM.W %D0-%D3/%A0-%A3, (%A5)
        TEST::SET_W(0, 0044220 | 5);
        TEST::SET_W(2, 0b0000111100001111);
        TEST::SET_W(6, TEST_BREAK);

        // MOVEM.W %D0-%D3/%A0-%A3, (-%A5)
        TEST::SET_W(8, 0044240 | 5);
        TEST::SET_W(10, 0b1111000011110000);
        TEST::SET_W(12, TEST_BREAK);

        // MOVEM.L %D0-%D3/%A0-%A3, (%A5)
        TEST::SET_W(14, 0044320 | 5);
        TEST::SET_W(16, 0b0000111100001111);
        TEST::SET_W(18, TEST_BREAK);

        // MOVEM.L %D0-%D3/%A0-%A3, (-%A5)
        TEST::SET_W(20, 0044340 | 5);
        TEST::SET_W(22, 0b1111000011110000);
        TEST::SET_W(24, TEST_BREAK);

        // MOVEM.W (%A5), %D0-%D3/%A0-%A3
        TEST::SET_W(26, 0046220 | 5);
        TEST::SET_W(28, 0b0000111100001111);
        TEST::SET_W(30, TEST_BREAK);

        // MOVEM.W (%A5+), %D0-%D3/%A0-%A3
        TEST::SET_W(32, 0046230 | 5);
        TEST::SET_W(34, 0b0000111100001111);
        TEST::SET_W(36, TEST_BREAK);

        // MOVEM.L (%A5), %D0-%D3/%A0-%A3
        TEST::SET_W(38, 0046320 | 5);
        TEST::SET_W(40, 0b0000111100001111);
        TEST::SET_W(42, TEST_BREAK);

        // MOVEM.L (%A5+), %D0-%D3/%A0-%A3
        TEST::SET_W(44, 0046330 | 5);
        TEST::SET_W(46, 0b0000111100001111);
        TEST::SET_W(48, TEST_BREAK);

        jit_compile(0, 50);
    }};
BOOST_FIXTURE_TEST_SUITE(MOVEM, Prepare, *boost::unit_test::fixture<F_MOVEM>())


BOOST_AUTO_TEST_SUITE(Store)
BOOST_AUTO_TEST_SUITE(Word)
BOOST_AUTO_TEST_SUITE(base)
BOOST_AUTO_TEST_CASE(ok) {
    cpu.D[0] = 0x0001;
    cpu.D[1] = 0x0002;
    cpu.D[2] = 0x0003;
    cpu.D[3] = 0x0004;
    cpu.A[0] = 0x1000;
    cpu.A[1] = 0x1001;
    cpu.A[2] = 0x1002;
    cpu.A[3] = 0x1003;
    cpu.A[5] = 0x200;
    run_test(0);
    BOOST_TEST(cpu.A[5] == 0x200);
    BOOST_TEST(TEST::GET_W(0x200) == 0x0001);
    BOOST_TEST(TEST::GET_W(0x202) == 0x0002);
    BOOST_TEST(TEST::GET_W(0x204) == 0x0003);
    BOOST_TEST(TEST::GET_W(0x206) == 0x0004);
    BOOST_TEST(TEST::GET_W(0x208) == 0x1000);
    BOOST_TEST(TEST::GET_W(0x20A) == 0x1001);
    BOOST_TEST(TEST::GET_W(0x20C) == 0x1002);
    BOOST_TEST(TEST::GET_W(0x20E) == 0x1003);
}

BOOST_AUTO_TEST_CASE(fail) {
    cpu.A[5] = 0xFFE;
	run_test(0);
	BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::AFAULT);
	BOOST_TEST(cpu.fault_SSW & SSW_CM);
	BOOST_TEST(cpu.ex_addr == 0x1000);
    BOOST_TEST(cpu.EA == 0xFFE);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(decr)
BOOST_AUTO_TEST_CASE(ok) {
    cpu.D[0] = 0x0001;
    cpu.D[1] = 0x0002;
    cpu.D[2] = 0x0003;
    cpu.D[3] = 0x0004;
    cpu.A[0] = 0x1000;
    cpu.A[1] = 0x1001;
    cpu.A[2] = 0x1002;
    cpu.A[3] = 0x1003;
    cpu.A[5] = 0x210;
    run_test(8);
    BOOST_TEST(TEST::GET_W(0x200) == 0x0001);
    BOOST_TEST(TEST::GET_W(0x202) == 0x0002);
    BOOST_TEST(TEST::GET_W(0x204) == 0x0003);
    BOOST_TEST(TEST::GET_W(0x206) == 0x0004);
    BOOST_TEST(TEST::GET_W(0x208) == 0x1000);
    BOOST_TEST(TEST::GET_W(0x20A) == 0x1001);
    BOOST_TEST(TEST::GET_W(0x20C) == 0x1002);
    BOOST_TEST(TEST::GET_W(0x20E) == 0x1003);
    BOOST_TEST(cpu.A[5] == 0x200);
}
BOOST_AUTO_TEST_CASE(fail) {
    cpu.A[5] = 0x4;
	run_test(8);
	BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::AFAULT);
	BOOST_TEST(cpu.fault_SSW & SSW_CM);
	BOOST_TEST(cpu.ex_addr == 0xfffffffe);
    BOOST_TEST(cpu.EA == 2);
}

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Long)
BOOST_AUTO_TEST_SUITE(base)

BOOST_AUTO_TEST_CASE(ok) {
    cpu.D[0] = 0x00000001;
    cpu.D[1] = 0x00000002;
    cpu.D[2] = 0x00000003;
    cpu.D[3] = 0x00000004;
    cpu.A[0] = 0x10000000;
    cpu.A[1] = 0x10000001;
    cpu.A[2] = 0x10000002;
    cpu.A[3] = 0x10000003;
    cpu.A[5] = 0x200;
    run_test(14);
    BOOST_TEST(TEST::GET_L(0x200) == 0x00000001);
    BOOST_TEST(TEST::GET_L(0x204) == 0x00000002);
    BOOST_TEST(TEST::GET_L(0x208) == 0x00000003);
    BOOST_TEST(TEST::GET_L(0x20C) == 0x00000004);
    BOOST_TEST(TEST::GET_L(0x210) == 0x10000000);
    BOOST_TEST(TEST::GET_L(0x214) == 0x10000001);
    BOOST_TEST(TEST::GET_L(0x218) == 0x10000002);
    BOOST_TEST(TEST::GET_L(0x21C) == 0x10000003);
    BOOST_TEST(cpu.A[5] == 0x200);
}
BOOST_AUTO_TEST_CASE(fail) {
    cpu.A[5] = 0xFFC;
	run_test(14);
	BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::AFAULT);
	BOOST_TEST(cpu.fault_SSW & SSW_CM);
	BOOST_TEST(cpu.ex_addr == 0x1000);
    BOOST_TEST(cpu.EA == 0xFFC);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(decr)

BOOST_AUTO_TEST_CASE(ok) {
    cpu.D[0] = 0x00000001;
    cpu.D[1] = 0x00000002;
    cpu.D[2] = 0x00000003;
    cpu.D[3] = 0x00000004;
    cpu.A[0] = 0x10000000;
    cpu.A[1] = 0x10000001;
    cpu.A[2] = 0x10000002;
    cpu.A[3] = 0x10000003;
    cpu.A[5] = 0x220;
    run_test(20);
    BOOST_TEST(TEST::GET_L(0x200) == 0x00000001);
    BOOST_TEST(TEST::GET_L(0x204) == 0x00000002);
    BOOST_TEST(TEST::GET_L(0x208) == 0x00000003);
    BOOST_TEST(TEST::GET_L(0x20C) == 0x00000004);
    BOOST_TEST(TEST::GET_L(0x210) == 0x10000000);
    BOOST_TEST(TEST::GET_L(0x214) == 0x10000001);
    BOOST_TEST(TEST::GET_L(0x218) == 0x10000002);
    BOOST_TEST(TEST::GET_L(0x21C) == 0x10000003);
    BOOST_TEST(cpu.A[5] == 0x200);
}
BOOST_AUTO_TEST_CASE(fail) {
    cpu.A[5] = 0x4;
	run_test(20);
	BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::AFAULT);
	BOOST_TEST(cpu.fault_SSW & SSW_CM);
	BOOST_TEST(cpu.ex_addr == 0xfffffffc);
    BOOST_TEST(cpu.EA == 0);
}

BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Load)
BOOST_AUTO_TEST_SUITE(Word)
BOOST_AUTO_TEST_SUITE(addr)

BOOST_AUTO_TEST_CASE(ok) {

    TEST::SET_W(0x200, 0x0001);
    TEST::SET_W(0x202, 0x0002);
    TEST::SET_W(0x204, 0x0003);
    TEST::SET_W(0x206, 0x0004);
    TEST::SET_W(0x208, 0x1000);
    TEST::SET_W(0x20A, 0x1001);
    TEST::SET_W(0x20C, 0x1002);
    TEST::SET_W(0x20E, 0x1003);
    cpu.A[5] = 0x200;
    run_test(26);
    BOOST_TEST(cpu.A[5] == 0x200);
    BOOST_TEST(cpu.D[0] == 0x0001);
    BOOST_TEST(cpu.D[1] == 0x0002);
    BOOST_TEST(cpu.D[2] == 0x0003);
    BOOST_TEST(cpu.D[3] == 0x0004);
    BOOST_TEST(cpu.A[0] == 0x1000);
    BOOST_TEST(cpu.A[1] == 0x1001);
    BOOST_TEST(cpu.A[2] == 0x1002);
    BOOST_TEST(cpu.A[3] == 0x1003);
}
BOOST_AUTO_TEST_CASE(fail) {
    cpu.A[5] = 0x2ffc;
	run_test(26);
	BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::AFAULT);
	BOOST_TEST(cpu.fault_SSW & SSW_CM);
	BOOST_TEST(cpu.ex_addr == 0x3000);
    BOOST_TEST(cpu.EA == 0x2ffc);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(incr)
BOOST_AUTO_TEST_CASE(ok) {

    TEST::SET_W(0x200, 0x0001);
    TEST::SET_W(0x202, 0x0002);
    TEST::SET_W(0x204, 0x0003);
    TEST::SET_W(0x206, 0x0004);
    TEST::SET_W(0x208, 0x1000);
    TEST::SET_W(0x20A, 0x1001);
    TEST::SET_W(0x20C, 0x1002);
    TEST::SET_W(0x20E, 0x1003);
    cpu.A[5] = 0x200;
    run_test(32);
    BOOST_TEST(cpu.A[5] == 0x210);
    BOOST_TEST(cpu.D[0] == 0x0001);
    BOOST_TEST(cpu.D[1] == 0x0002);
    BOOST_TEST(cpu.D[2] == 0x0003);
    BOOST_TEST(cpu.D[3] == 0x0004);
    BOOST_TEST(cpu.A[0] == 0x1000);
    BOOST_TEST(cpu.A[1] == 0x1001);
    BOOST_TEST(cpu.A[2] == 0x1002);
    BOOST_TEST(cpu.A[3] == 0x1003);
}
BOOST_AUTO_TEST_CASE(fail) {
    cpu.A[5] = 0x2ffc;
	run_test(32);
	BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::AFAULT);
	BOOST_TEST(cpu.fault_SSW & SSW_CM);
	BOOST_TEST(cpu.ex_addr == 0x3000);
    BOOST_TEST(cpu.EA == 0x2ffc);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Long)

BOOST_AUTO_TEST_SUITE(addr)

BOOST_AUTO_TEST_CASE(ok) {

    TEST::SET_L(0x200, 0x00000001);
    TEST::SET_L(0x204, 0x00000002);
    TEST::SET_L(0x208, 0x00000003);
    TEST::SET_L(0x20C, 0x00000004);
    TEST::SET_L(0x210, 0x10000000);
    TEST::SET_L(0x214, 0x10000001);
    TEST::SET_L(0x218, 0x10000002);
    TEST::SET_L(0x21C, 0x10000003);
    cpu.A[5] = 0x200;
    run_test(38);
    BOOST_TEST(cpu.A[5] == 0x200);
    BOOST_TEST(cpu.D[0] == 0x00000001);
    BOOST_TEST(cpu.D[1] == 0x00000002);
    BOOST_TEST(cpu.D[2] == 0x00000003);
    BOOST_TEST(cpu.D[3] == 0x00000004);
    BOOST_TEST(cpu.A[0] == 0x10000000);
    BOOST_TEST(cpu.A[1] == 0x10000001);
    BOOST_TEST(cpu.A[2] == 0x10000002);
    BOOST_TEST(cpu.A[3] == 0x10000003);
}
BOOST_AUTO_TEST_CASE(fail) {
    cpu.A[5] = 0x2ffc;
	run_test(38);
	BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::AFAULT);
	BOOST_TEST(cpu.fault_SSW & SSW_CM);
	BOOST_TEST(cpu.ex_addr == 0x3000);
    BOOST_TEST(cpu.EA == 0x2ffc);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(incr)
BOOST_AUTO_TEST_CASE(ok) {

    TEST::SET_L(0x200, 0x00000001);
    TEST::SET_L(0x204, 0x00000002);
    TEST::SET_L(0x208, 0x00000003);
    TEST::SET_L(0x20C, 0x00000004);
    TEST::SET_L(0x210, 0x10000000);
    TEST::SET_L(0x214, 0x10000001);
    TEST::SET_L(0x218, 0x10000002);
    TEST::SET_L(0x21C, 0x10000003);
    cpu.A[5] = 0x200;
    run_test(44);
    BOOST_TEST(cpu.A[5] == 0x220);
    BOOST_TEST(cpu.D[0] == 0x00000001);
    BOOST_TEST(cpu.D[1] == 0x00000002);
    BOOST_TEST(cpu.D[2] == 0x00000003);
    BOOST_TEST(cpu.D[3] == 0x00000004);
    BOOST_TEST(cpu.A[0] == 0x10000000);
    BOOST_TEST(cpu.A[1] == 0x10000001);
    BOOST_TEST(cpu.A[2] == 0x10000002);
    BOOST_TEST(cpu.A[3] == 0x10000003);
}
BOOST_AUTO_TEST_CASE(fail) {
    cpu.A[5] = 0x2ffc;
	run_test(44);
	BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::AFAULT);
	BOOST_TEST(cpu.fault_SSW & SSW_CM);
	BOOST_TEST(cpu.ex_addr == 0x3000);
    BOOST_TEST(cpu.EA == 0x2ffc);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/movem.lua
-- MOVEM.W %D0-%D3/%A0-%A3, (%A5)
MEM[0].W = OCT "044220" | 5
MEM[2].W = tonumber("0000111100001111", 2)
MEM[4].W = TEST.BREAK

-- MOVEM.W %D0-%D3/%A0-%A3, -(%A5)
MEM[6].W = OCT "044240" | 5
MEM[8].W = tonumber("1111000011110000", 2)
MEM[10].W = TEST.BREAK

-- MOVEM.W (%A5), %D0-%D3/%A0-%A3
MEM[12].W = OCT "046220" | 5
MEM[14].W = tonumber("0000111100001111", 2)
MEM[16].W = TEST.BREAK

-- MOVEM.W (%A5)+, %D0-%D3/%A0-%A3
MEM[18].W = OCT "046230" | 5
MEM[20].W = tonumber("0000111100001111", 2)
MEM[22].W = TEST.BREAK

-- MOVEM.L %D0-%D3/%A0-%A3, (%A6)
MEM[24].W = OCT "044320" | 6
MEM[26].W = tonumber("0000111100001111", 2)
MEM[28].W = TEST.BREAK

-- MOVEM.L %D0-%D3/%A0-%A3, -(%A6)
MEM[30].W = OCT "044340" | 6
MEM[32].W = tonumber("1111000011110000", 2)
MEM[34].W = TEST.BREAK

-- MOVEM.L (%A6), %D0-%D3/%A0-%A3
MEM[36].W = OCT "046320" | 6
MEM[38].W = tonumber("0000111100001111", 2)
MEM[40].W = TEST.BREAK

-- MOVEM.L (%A6)+, %D0-%D3/%A0-%A3
MEM[42].W = OCT "046330" | 6
MEM[44].W = tonumber("0000111100001111", 2)
MEM[46].W = TEST.BREAK

JIT_COMPILE(0, 48)

local run = {}
run.W = {
    store = {
        base = {
            ok = function()
                CPU.D[1].W = 0x0001
                CPU.D[2].W = 0x0002
                CPU.D[3].W = 0x0003
                CPU.D[4].W = 0x0004
                CPU.A[1] = 0x1001
                CPU.A[2] = 0x1002
                CPU.A[3] = 0x1003
                CPU.A[4] = 0x1004
                CPU.A[6] = 0x1200
                TEST.RUN(0)
                TEST.CHECK("CPU.A[6]", 0x1200)
                TEST.CHECK("MEM[0x1200].W", 0x0001)
                TEST.CHECK("MEM[0x1202].W", 0x0002)
                TEST.CHECK("MEM[0x1204].W", 0x0003)
                TEST.CHECK("MEM[0x1206].W", 0x0004)
                TEST.CHECK("MEM[0x1208].W", 0x1001)
                TEST.CHECK("MEM[0x120A].W", 0x1002)
                TEST.CHECK("MEM[0x120C].W", 0x1003)
                TEST.CHECK("MEM[0x120E].W", 0x1004)
            end,
            fail = function()
                CPU.A[6] = 0x1FFE
                TEST.RUN(0, "access fault")
                TEST.CHECK("CPU.SSW.CM", true)
                TEST.CHECK("CPU.ex_addr", 0x2000)
                TEST.CHECK("CPU.EA", 0x1FFE)
            end
        },
        decr = {
            ok = function()
                CPU.D[1].W = 0x0001
                CPU.D[2].W = 0x0002
                CPU.D[3].W = 0x0003
                CPU.D[4].W = 0x0004
                CPU.A[1] = 0x1001
                CPU.A[2] = 0x1002
                CPU.A[3] = 0x1003
                CPU.A[4] = 0x1004
                CPU.A[6] = 0x1210
                TEST.RUN(6)
                TEST.CHECK("CPU.A[6]", 0x1200)
                TEST.CHECK("MEM[0x1200].W", 0x0001)
                TEST.CHECK("MEM[0x1202].W", 0x0002)
                TEST.CHECK("MEM[0x1204].W", 0x0003)
                TEST.CHECK("MEM[0x1206].W", 0x0004)
                TEST.CHECK("MEM[0x1208].W", 0x1001)
                TEST.CHECK("MEM[0x120A].W", 0x1002)
                TEST.CHECK("MEM[0x120C].W", 0x1003)
                TEST.CHECK("MEM[0x120E].W", 0x1004)
            end,
            fail = function()
                CPU.A[6] = 0x1004
                TEST.RUN(6, "access fault")
                TEST.CHECK("CPU.SSW.CM", true)
                TEST.CHECK("CPU.ex_addr", 0xFFE)
                TEST.CHECK("CPU.EA", 0x1002)
            end
        }
    },
    load = {
        base = {
            ok = function()
                MEM[0x1200].W = 0x0001
                MEM[0x1202].W = 0x0002
                MEM[0x1204].W = 0x0003
                MEM[0x1206].W = 0x0004
                MEM[0x1208].W = 0x1000
                MEM[0x120A].W = 0x1001
                MEM[0x120C].W = 0x1002
                MEM[0x120E].W = 0x1003
                CPU.A[6] = 0x1200
                TEST.RUN(12)
                TEST.CHECK("CPU.A[6]", 0x1200)
                TEST.CHECK("CPU.D[1].W", 0x0001)
                TEST.CHECK("CPU.D[2].W", 0x0002)
                TEST.CHECK("CPU.D[3].W", 0x0003)
                TEST.CHECK("CPU.D[4].W", 0x0004)
                TEST.CHECK("CPU.A[1]", 0x1000)
                TEST.CHECK("CPU.A[2]", 0x1001)
                TEST.CHECK("CPU.A[3]", 0x1002)
                TEST.CHECK("CPU.A[4]", 0x1003)
            end,
            fail = function()
                CPU.A[6] = 0x2FFC
                TEST.RUN(12, "access fault")
                TEST.CHECK("CPU.SSW.CM", true)
                TEST.CHECK("CPU.ex_addr", 0x3000)
                TEST.CHECK("CPU.EA", 0x2FFC)
            end
        },
        incr = {
            ok = function()
                MEM[0x1200].W = 0x0001
                MEM[0x1202].W = 0x0002
                MEM[0x1204].W = 0x0003
                MEM[0x1206].W = 0x0004
                MEM[0x1208].W = 0x1000
                MEM[0x120A].W = 0x1001
                MEM[0x120C].W = 0x1002
                MEM[0x120E].W = 0x1003
                CPU.A[6] = 0x1200
                TEST.RUN(18)
                TEST.CHECK("CPU.A[6]", 0x1210)
                TEST.CHECK("CPU.D[1].W", 0x0001)
                TEST.CHECK("CPU.D[2].W", 0x0002)
                TEST.CHECK("CPU.D[3].W", 0x0003)
                TEST.CHECK("CPU.D[4].W", 0x0004)
                TEST.CHECK("CPU.A[1]", 0x1000)
                TEST.CHECK("CPU.A[2]", 0x1001)
                TEST.CHECK("CPU.A[3]", 0x1002)
                TEST.CHECK("CPU.A[4]", 0x1003)
            end,
            fail = function()
                CPU.A[6] = 0x2FFC
                TEST.RUN(18, "access fault")
                TEST.CHECK("CPU.SSW.CM", true)
                TEST.CHECK("CPU.ex_addr", 0x3000)
                TEST.CHECK("CPU.EA", 0x2FFC)
            end
        }
    }
}
run.L = {
    store = {
        base = {
            ok = function()
                CPU.D[1].L = 0x00000001
                CPU.D[2].L = 0x00000002
                CPU.D[3].L = 0x00000003
                CPU.D[4].L = 0x00000004
                CPU.A[1] = 0x10000001
                CPU.A[2] = 0x10000002
                CPU.A[3] = 0x10000003
                CPU.A[4] = 0x10000004
                CPU.A[7] = 0x1200
                TEST.RUN(24)
                TEST.CHECK("CPU.A[7]", 0x1200)
                TEST.CHECK("MEM[0x1200].L", 0x00000001)
                TEST.CHECK("MEM[0x1204].L", 0x00000002)
                TEST.CHECK("MEM[0x1208].L", 0x00000003)
                TEST.CHECK("MEM[0x120C].L", 0x00000004)
                TEST.CHECK("MEM[0x1210].L", 0x10000001)
                TEST.CHECK("MEM[0x1214].L", 0x10000002)
                TEST.CHECK("MEM[0x1218].L", 0x10000003)
                TEST.CHECK("MEM[0x121C].L", 0x10000004)
            end,
            fail = function()
                CPU.A[7] = 0x1FFC
                TEST.RUN(24, "access fault")
                TEST.CHECK("CPU.SSW.CM", true)
                TEST.CHECK("CPU.ex_addr", 0x2000)
                TEST.CHECK("CPU.EA", 0x1FFC)
            end
        },
        decr = {
            ok = function()
                CPU.D[1].L = 0x00000001
                CPU.D[2].L = 0x00000002
                CPU.D[3].L = 0x00000003
                CPU.D[4].L = 0x00000004
                CPU.A[1] = 0x10000001
                CPU.A[2] = 0x10000002
                CPU.A[3] = 0x10000003
                CPU.A[4] = 0x10000004
                CPU.A[7] = 0x1220
                TEST.RUN(30)
                TEST.CHECK("CPU.A[7]", 0x1200)
                TEST.CHECK("MEM[0x1200].L", 0x00000001)
                TEST.CHECK("MEM[0x1204].L", 0x00000002)
                TEST.CHECK("MEM[0x1208].L", 0x00000003)
                TEST.CHECK("MEM[0x120C].L", 0x00000004)
                TEST.CHECK("MEM[0x1210].L", 0x10000001)
                TEST.CHECK("MEM[0x1214].L", 0x10000002)
                TEST.CHECK("MEM[0x1218].L", 0x10000003)
                TEST.CHECK("MEM[0x121C].L", 0x10000004)
            end,
            fail = function()
                CPU.A[7] = 0x1004
                TEST.RUN(30, "access fault")
                TEST.CHECK("CPU.SSW.CM", true)
                TEST.CHECK("CPU.ex_addr", 0xFFC)
                TEST.CHECK("CPU.EA", 0x1000)
            end
        }
    },
    load = {
        base = {
            ok = function()
                MEM[0x1200].L = 0x00000001
                MEM[0x1204].L = 0x00000002
                MEM[0x1208].L = 0x00000003
                MEM[0x120C].L = 0x00000004
                MEM[0x1210].L = 0x10000000
                MEM[0x1214].L = 0x10000001
                MEM[0x1218].L = 0x10000002
                MEM[0x121C].L = 0x10000003
                CPU.A[7] = 0x1200
                TEST.RUN(36)
                TEST.CHECK("CPU.A[7]", 0x1200)
                TEST.CHECK("CPU.D[1].L", 0x00000001)
                TEST.CHECK("CPU.D[2].L", 0x00000002)
                TEST.CHECK("CPU.D[3].L", 0x00000003)
                TEST.CHECK("CPU.D[4].L", 0x00000004)
                TEST.CHECK("CPU.A[1]", 0x10000000)
                TEST.CHECK("CPU.A[2]", 0x10000001)
                TEST.CHECK("CPU.A[3]", 0x10000002)
                TEST.CHECK("CPU.A[4]", 0x10000003)
            end,
            fail = function()
                CPU.A[7] = 0x2FFC
                TEST.RUN(36, "access fault")
                TEST.CHECK("CPU.SSW.CM", true)
                TEST.CHECK("CPU.ex_addr", 0x3000)
                TEST.CHECK("CPU.EA", 0x2FFC)
            end
        },
        incr = {
            ok = function()
                MEM[0x1200].L = 0x00000001
                MEM[0x1204].L = 0x00000002
                MEM[0x1208].L = 0x00000003
                MEM[0x120C].L = 0x00000004
                MEM[0x1210].L = 0x10000000
                MEM[0x1214].L = 0x10000001
                MEM[0x1218].L = 0x10000002
                MEM[0x121C].L = 0x10000003
                CPU.A[7] = 0x1200
                TEST.RUN(42)
                TEST.CHECK("CPU.A[7]", 0x1220)
                TEST.CHECK("CPU.D[1].L", 0x00000001)
                TEST.CHECK("CPU.D[2].L", 0x00000002)
                TEST.CHECK("CPU.D[3].L", 0x00000003)
                TEST.CHECK("CPU.D[4].L", 0x00000004)
                TEST.CHECK("CPU.A[1]", 0x10000000)
                TEST.CHECK("CPU.A[2]", 0x10000001)
                TEST.CHECK("CPU.A[3]", 0x10000002)
                TEST.CHECK("CPU.A[4]", 0x10000003)
            end,
            fail = function()
                CPU.A[7] = 0x2FFC
                TEST.RUN(42, "access fault")
                TEST.CHECK("CPU.SSW.CM", true)
                TEST.CHECK("CPU.ex_addr", 0x3000)
                TEST.CHECK("CPU.EA", 0x2FFC)
            end
        }
    }
}
return run
